Uncaught Error: Cannot find module 'eslint-plugin-react'

[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 0.196.0
**System**: Mac OS X 10.10.2
**Thrown From**: [linter-eslint](https://github.com/AtomLinter/linter-eslint) package, v0.5.3

### Stack Trace

Uncaught Error: Cannot find module 'eslint-plugin-react'

```
At module.js:330

Error: Cannot find module 'eslint-plugin-react'
  at Module._resolveFilename (module.js:328:15)
  at Function.Module._resolveFilename (/Applications/Atom.app/Contents/Resources/app.asar/src/module-cache.js:383:52)
  at Function.Module._load (module.js:270:25)
  at Module.require (module.js:357:17)
  at require (module.js:376:17)
  at /Users/kaelan/www/laconic/node_modules/eslint/lib/cli-engine.js:107:26
  at Array.forEach (native)
  at loadPlugins (/Users/kaelan/www/laconic/node_modules/eslint/lib/cli-engine.js:98:21)
  at new CLIEngine (/Users/kaelan/www/laconic/node_modules/eslint/lib/cli-engine.js:268:5)
  at LinterESLint.lintFile (/Users/kaelan/.atom/packages/linter-eslint/lib/linter-eslint.coffee:73:22)
  at /Users/kaelan/.atom/packages/linter/lib/linter-view.coffee:163:18
  at Array.forEach (native)
  at /Users/kaelan/.atom/packages/linter/lib/linter-view.coffee:162:18
  at FSReqWrap.oncomplete (fs.js:77:15)

```

### Commands

```
     -0:03.1.0 core:save (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "disabledPackages": [
      "terminal",
      "tree-view-open-files",
      "linter-csslint",
      "linter-scss-lint",
      "linter-jscs",
      "eslint",
      "linter-jshint"
    ],
    "themes": [
      "flatland-dark-ui",
      "flatland-dark"
    ]
  },
  "linter-eslint": {
    "disableWhenNoEslintrcFileInPath": true
  }
}
```

### Installed Packages

```coffee
# User
angularjs, v0.1.0
auto-detect-indentation, v0.4.0
autocomplete-plus, v2.12.1
autocomplete-snippets, v1.3.0
color-picker, v1.7.0
cursor-count, v0.2.0
emmet, v2.3.8
file-icons, v1.5.4
flatland-dark, v0.1.6
flatland-dark-ui, v0.2.4
highlight-selected, v0.9.2
less-than-slash, v0.6.0
linter, v0.12.1
linter-coffeelint, v0.2.2
linter-eslint, v0.5.3
linter-less, v0.3.1
linter-pylint, v0.2.1
react, v0.11.7
white-cursor, v0.6.0

# Dev
No dev packages
```

--- a/lib/linter-eslint.coffee
+++ b/lib/linter-eslint.coffee
@@ -25,7 +25,7 @@ class LinterESLint extends Linter
       eslint = require(eslintPath)
       @localEslint = true
       return eslint
-    # Fall back to the version packaged in linster-eslint
+    # Fall back to the version packaged in linter-eslint
     return require('eslint')
 
   lintFile: (filePath, callback) ->
@@ -60,17 +60,34 @@ class LinterESLint extends Linter
 
     config = engine.getConfigForFile(origPath)
 
-    # This sidesteps a chicken and egg problem:
-    # CLIEngine contains loadPlugins() that is not exposed, so we can't call it
-    # directly. The plugins need to be passed into CLIEngine, but we don't know
-    # which plugins are loaded until after we load CLIEngine.
-    #
-    # If you are loading plugins this will replace the existing engine with a
-    # new engine where we can pass in the set of plugins for it to load.
+    notFullyCompatible = false
+    notFoundPlugins = []
     if config.plugins?.length
       if @localEslint
-        options.plugins = config.plugins
-        engine = new CLIEngine(options)
+        unless engine.addPlugin
+          # we have an old version `eslint@0.20` or less
+          notFullyCompatible = true
+          # This sidesteps a chicken and egg problem:
+          # CLIEngine contains loadPlugins() that is not exposed, so we can't call it
+          # directly. The plugins need to be passed into CLIEngine, but we don't know
+          # which plugins are loaded until after we load CLIEngine.
+          #
+          # If you are loading plugins this will replace the existing engine with a
+          # new engine where we can pass in the set of plugins for it to load.
+          options.plugins = config.plugins
+          engine = new CLIEngine(options)
+        else
+          # we have `eslint@0.21`+
+          config.plugins.forEach (pluginName) ->
+            npmPluginName = 'eslint-plugin-' + pluginName
+            try
+              pluginPath = resolve(npmPluginName, {
+                basedir: path.dirname(origPath)
+              })
+              pluginObject = require(pluginPath)
+              engine.addPlugin(npmPluginName, pluginObject)
+            catch
+              notFoundPlugins.push(npmPluginName)
       else
         isPluginRule = new RegExp("^(#{config.plugins.join('|')})/")
         Object.keys(config.rules).forEach (key) ->
@@ -79,9 +96,18 @@ class LinterESLint extends Linter
     # wrap `eslint()` into `allowUnsafeNewFunction`
     # https://discuss.atom.io/t/--template-causes-unsafe-eval-error/9310
     # https://github.com/babel/babel/blob/master/src/acorn/src/identifier.js#L46
-    result = null
-    allowUnsafeNewFunction =>
-      result = linter.verify @editor.getText(), config
+    result = []
+    if notFoundPlugins.length
+      result.push({
+        line: 1
+        column: 0
+        severity: 1
+        message: "`npm install #{notFoundPlugins.join(' ')}`
+        in your project (linter-eslint)"
+      })
+    else
+      allowUnsafeNewFunction =>
+        result = linter.verify @editor.getText(), config
 
     if config.plugins?.length and not @localEslint
       result.push({
@@ -90,6 +116,15 @@ class LinterESLint extends Linter
         severity: 1
         message: "`npm install eslint` in your project to enable plugins:
         #{config.plugins.join(', ')} (linter-eslint)"
+      })
+
+    if notFullyCompatible
+      result.push({
+        line: 1
+        column: 0
+        severity: 1
+        message: "You are using the version of eslint@0.20 or less.
+        You have to update eslint to 0.21+ or downgrade linter-eslint"
       })
 
     messages = result.map (m) =>

